tests/test/constants.rs

```diff
@@ -105,6 +122,99 @@ fn generic_impl() {
     }
 }
 
+#[test]
+fn unevaluated_impl() {
+    test! {
+        program {
+            struct S<const N> {}
+
+            trait Trait {}
+
+            impl Trait for S<3?> {}
+        }
+
+        goal {
+            exists<const N> {
+                S<N>: Trait
+            }
+        } yields {
+            "Unique; substitution [?0 := 3?], lifetime constraints []"
+        }
+
+        goal {
+            forall<const N> {
+                S<N>: Trait
+            }
+        } yields {
+            "No possible solution"
+        }
+
+        goal {
+            S<8>: Trait
+        } yields {
+            "No possible solution"
+        }
+
+        goal {
+            S<3>: Trait
+        } yields {
+            "Unique"
+        }
+
+        goal {
+            S<?>: Trait
+        } yields {
+            "Ambiguous; no inference guidance"
+        }
+
+        goal {
+            S<3?>: Trait
+        } yields {
+            "Unique"
+        }
+
+        goal {
+            S<8?>: Trait
+        } yields {
+            "No possible solution"
+        }
+    }
+}
+
+#[test]
+fn unevaluated_impl_generic() {
+    test! {
+        program {
+            struct S<const N> {}
+
+            trait Trait {}
+
+            impl Trait for S<?> {}
+        }
+
+        goal {
+            S<0>: Trait
+        } yields {
+            "Ambiguous; no inference guidance"
+        }
+
+        goal {
+            S<?>: Trait
+        } yields {
+            "Unique"
```

Comment:
See, this should be `Ambiguous`, for a similar reason that two `InferenceVar`s aren't the same: we don't actually know this would evaluate to the same const.

Reply:
I've been assuming that for chalk testing purposes that `?` evaluates to a specific but unknown constant. I don't feel like breaking reflexivity.

Reply:
But that's not true. The point here is that `?` represents a constant that can't be evaluated *yet*. Importantly, if you have two constants that can't be evaluated (e.g. `{N+1}` and `{X +1}`), then we can't say that if there is a solution that covers one case, there is a solution that covers all cases.

Reply:
As it turns out, removing `const_eq` for unevaluated consts and assuming that two `?`s don't necessarily evaluate to the same value breaks `LocalImplAllowed(Baz: Foo<?>)` in the `unevaluated_const_unknown` coherence test.

Reply:
Oops, I never responded to this. Sorry about that.

So, I think theoretically we could always say that `impl Foo<?0> for Baz {}` never violates orphan rules, since both `Foo` and `Baz` are local. *However*, I think for unevaluated consts, it's okay to say it's `Ambiguous` if we get `TooGeneric`. I would have to look more into the orphan rules code to figure out how to best handle this. For this PR, we can mark a comment or `FIXME` and file a followup issue for more discussion.